src/dotty/tools/dotc/transform/ExtensionMethods.scala

```diff
@@ -181,4 +205,63 @@ class ExtensionMethods extends MiniPhaseTransform with DenotTransformer with Ful
       cpy.DefDef(tree)(rhs = atGroupEnd(forwarder(extensionMeth, tree)(_)))
     } else tree
   }
+
+  /** Implement Step 2 of SIP-15: rerouting calls.
+   *  Note that this is different from Scala 2 which did the rerouting much later in
+   *  the PostErasure phase.
+   *  TODO: avoid code duplication with fullyParameterizedDef#rewireTree
+   */
+  private def rewire(select: Select, targs: List[Tree])(implicit ctx: Context): Tree = {
+    val Select(qual, _) = select
+    val origMeth = select.symbol
+    if (isMethodWithExtension(origMeth)) {
+      val extensionMeth = extensionMethod(origMeth)
+      val origClass = origMeth.enclosingClass.asClass
+      val base = qual.tpe.baseTypeWithArgs(origClass)
+      assert(base.exists)
+      ref(extensionMeth)
+        .appliedToTypeTrees(targs ++ base.argInfos.map(TypeTree(_)))
+        .appliedTo(qual)
+    } else EmptyTree
+  }
+
+  override def transformTypeApply(tree: TypeApply)(implicit ctx: Context, info: TransformerInfo): Tree =
+    tree match {
+      case TypeApply(sel @ Select(_,_), args) =>
+        rewire(sel, args) orElse tree
+      case _ =>
+        tree
+    }
+
+  override def transformSelect(tree: Select)(implicit ctx: Context, info: TransformerInfo): Tree =
+    tree.tpe.widen match {
+      case tp: PolyType =>
+        tree // The rewiring will be handled by transformTypeApply
+      case tp =>
+        rewire(tree, Nil) orElse tree
+    }
+
+
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo): Tree =
+    tree match {
+      // There is no extension method for == and != because they're defined in Any,
+      // but we still want to replace them by calls to == and != on the unboxed values
+      // to take advantage of the optimizations done by PeepholeOptimization, so we do
+      // the following transform here:
+      // v1 == v2 => v1.u == v2.u
+      // v1 != v2 => v1.u != v2.u
+      case BinaryOp(v1, op, v2) if (op eq defn.Any_==) || (op eq defn.Any_!=) =>
+        val tpw1 = v1.tpe.widen
+        val tpw2 = v2.tpe.widen
+        val sym1 = tpw1.typeSymbol
+        if (isDerivedValueClass(sym1) && (tpw1 =:= tpw2)) {
```

Comment:
Probably yes, I used =:= because that's how it was done in scalac: https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/ast/TreeInfo.scala#L71 I'll try just using `eq` and see if it breaks anything.